search-provider/eks-discovery-feed-provider.c

```diff
@@ -359,11 +391,151 @@ handle_article_card_descriptions (EksDiscoveryFeedDatabaseContentProvider *skele
 
     return TRUE;
 }
-                       
+
+static void
+recent_news_articles_cb (GObject *source,
+                         GAsyncResult *result,
+                         gpointer user_data)
+{
+  EkncEngine *engine = EKNC_ENGINE (source);
+  DiscoveryFeedQueryState *state = user_data;
+
+  g_application_release (g_application_get_default ());
+
+  GError *error = NULL;
+  EkncDomain *domain = eknc_engine_get_domain_for_app (engine, state->provider->application_id, &error);
+  if (domain == NULL)
+    {
+      g_dbus_method_invocation_take_error (state->invocation, error);
+      discovery_feed_query_state_free (state);
+      return;
+    }
+
+  GSList *shards = eknc_domain_get_shards (domain);
+
+  GVariantBuilder shard_builder;
+  g_variant_builder_init (&shard_builder, G_VARIANT_TYPE_STRING_ARRAY);
+  for (GSList *l = shards; l; l = l->next)
+    {
+      g_autofree gchar *shard_path = NULL;
+      EosShardShardFile *shard = l->data;
+
+      g_object_get (shard, "path", &shard_path, NULL);
+      g_variant_builder_add (&shard_builder, "s", shard_path);
+    }
+
+  g_autoptr(EkncQueryResults) results = NULL;
+  if (!(results = eknc_engine_query_finish (engine, result, &error)))
+    {
+      g_dbus_method_invocation_return_gerror (state->invocation, error);
+      discovery_feed_query_state_free (state);
+      return;
+    }
+
+  GSList *models = eknc_query_results_get_models (results);
+
+  GVariantBuilder builder;
+  g_variant_builder_init (&builder, G_VARIANT_TYPE ("aa{ss}"));
+  for (GSList *l = models; l; l = l->next)
+    {
+      /* Start building up object */
+      g_variant_builder_open (&builder, G_VARIANT_TYPE ("a{ss}"));
+
+      EkncContentObjectModel *model = l->data;
+      DiscoveryFeedCustomProps flags = 0;
+
+      /* Add key-value pairs based on things we haven't addded yet */
+      if (!(flags & DISCOVERY_FEED_SET_CUSTOM_TITLE))
+        {
+          add_key_value_pair_from_model_to_variant (model, &builder, "title");
+          add_key_value_pair_from_model_to_variant (model, &builder, "synopsis");
+        }
+
+      add_key_value_pair_from_model_to_variant (model, &builder, "last-modified-date");
+      add_key_value_pair_from_model_to_variant (model, &builder, "thumbnail-uri");
+      add_key_value_pair_from_model_to_variant (model, &builder, "ekn-id");
+
+      /* Stop building object */
+      g_variant_builder_close (&builder);
+    }
+  g_dbus_method_invocation_return_value (state->invocation,
+                                         g_variant_new ("(asaa{ss})", &shard_builder, &builder));
+  discovery_feed_query_state_free (state);
+}
+
+static gboolean
+handle_get_recent_news (EksDiscoveryFeedDatabaseContentProvider *skeleton,
+                        GDBusMethodInvocation                  *invocation,
+                        gpointer                                user_data)
+{
+    EksDiscoveryFeedDatabaseContentProvider *self = (EksDiscoveryFeedDatabaseContentProvider *) user_data;
```

Comment:
Either use `EKS_DISCOVERY_FEED_DATABASE_CONTENT_PROVIDER()` to cast, or no cast at all, since `user_data` is `void *`